dev/camdenhells/REQUIREMENTS.md

```diff
@@ -324,41 +350,75 @@ This does not preclude performance impacts; clearly there will be limits, but ha
   * Isolated - from all (other) render thread/event processing
     * Frame draws, window re-sizes, queued events, other CRUD transactions, etc.
     * Implies running on the render thread as an event, queued from any thread
-    * TODO how to handle queued transactions after window detach/loss of context...
+    * Queued transactions after window detach/loss of context are dropped
   * Durable - process persistence scope
     * Combined state preserved across Activities and render contexts
     * Essentially a requirement to provide automatic restores
 3. Provide "successful completion" callback
-  * Provide a variant that includes queueing to main/UI thread ? 
-
-# 4. Testing, Debugging, and Reliability Requirements
-
-1. Automated unit/integration test harness and tests 
-  * Details ?
-2. Internal runtime logging/tracing/assertions?
-  * E.g. for integration testing and for debugging support issues
-3. Compile-time API-usage validation, e.g. annotations ?
-  * Are these enforceable e.g. via lint?
-4. Runtime API-usage validation for the app, e.g. debug-only assertions/exceptions?
-  * 
-5. Debug visualizations (#1436)
-  * Trident, camera eyeball/frustum, lighting rays/lines, object bounds
-  * Are there (non-debug) app use cases for some of these as well? e.g.: 
-    * Interactive camera/lighting selections from a global-view PIP inset 
-    * Theater/stage lighting design apps
-    * Interior design apps - door/window/furniture/light placements
-  * If so perhaps just call them e.g. camera/light/bounds visualizations  
+  * Provide a variant that includes queueing to main/UI thread (see above)
+
+# 4. Testing, Validation, and Debugging Requirements
+
+## 4.1 Compile-time
+
+  * Consistently specify any constraints on all member and parameter values
+    * Specify all references as either nullable or non-null
+    * Specify integer and float ranges whenever less than built-in language limits
+  * ~Validate public API-usage [2.1]~
+     * ~Create an Android lint plugin to detect invocations of non API methods?~
+     * Seems like a reliable library-specific tool would be a lot of work/maintenance, and unlikely that client developers would download/install/use it anyway
+
+## 4.2 Build-time 
+
+  Help! I'm trying to define the line on what should be unit tested and what should not, and don't even understand the hardware/software environment in which unit tests run (i.e. emulators use what version of OpenGL? is there an Android process/activity context? etc)
+  
+  * Automated unit/integration test harness and tests 
+  * For all public API and internal interface methods except those:
+    * With significant historical runtime event dependencies or behavioral states, and/or
+    * Whose success cannot be (at least partially) reasonably determined as value assertions
+  * Includes e.g. contructors, simple get/set accessors, deterministic functions/transforms
+  * Excludes e.g. frame draws, client callbacks
```

Comment:
My ultimate goal is that the examples can have full integration testing where screenshots can be captured and compared to known good screenshots from the same device. This will represent a full stack integration test that should help us catch when examples break as well as assert that low level changes did not have unforeseen larger scale effects such as what #1986 is resolving.

Reply:
WIll add examples integration tests (as [2.1] unless you think [2.0]). Should we mention instrumentation? Seems like we are currently not using it, but maybe I'm missing something in my analysis.  There is also a new Android Test Orchestrator that looks like it might be useful for this...  All new to me, just exploring the edges for requirements purposes for now.

Reply:
2.1 is fine for now. I will look into the Test Orchestrator